From 01780f58e1b81b2584fb9466e63ae8c9d486c608 Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <thomas.langer@intel.com>
Date: Fri, 3 Jul 2020 12:55:26 +0200
Subject: [PATCH] PONRTSYS-7614: xpcs: Allow to apply current settings

don't return an error if the already active mode is configured again.
Only skip the xpcs reinit for this case.
---
 drivers/net/ethernet/lantiq/xpcs/xpcs.c | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs.c b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
index 841b3526d894..b17f77c2807e 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs.c
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
@@ -1557,24 +1557,26 @@ static int xpcs_ethtool_ksettings_set(struct device *dev,
 	    speed != SPEED_1000 &&
 	    speed != SPEED_100 &&
 	    speed != SPEED_10 &&
-	    speed != SPEED_2500)
+	    speed != SPEED_2500) {
+		dev_err(dev, "XPCS invalid speed %d\n", speed);
 		return -EINVAL;
+	}
 
-	if ((speed == SPEED_1000 || speed == SPEED_100 || speed == SPEED_10)
-	    && ((pdata->conntype == CONN_TYPE_ANEG_DIS && pdata->speed != speed)
-		|| pdata->mode != ONEG_XAUI_MODE)) {
+	if ((speed == SPEED_1000 || speed == SPEED_100 || speed == SPEED_10) &&
+	    (pdata->conntype == CONN_TYPE_ANEG_DIS && pdata->speed != speed))
 		mode = ONEG_XAUI_MODE;
-	} else if (speed == SPEED_10000 && (pdata->mode != TENG_KR_MODE))
+	else if (speed == SPEED_10000)
 		mode = TENG_KR_MODE;
-	else if (speed == SPEED_2500 && (pdata->mode != TWOP5G_SGMII_MODE))
+	else if (speed == SPEED_2500)
 		mode = TWOP5G_SGMII_MODE;
 	else
-		return -1;
+		return -EINVAL;
 
 	pdata->speed = speed;
 
 	/* Restart Xpcs & PHY */
-	xpcs_reinit(pdata->dev, mode);
+	if (mode != pdata->mode)
+		xpcs_reinit(pdata->dev, mode);
 
 	return 0;
 }
